tket2/src/circuit/command.rs

```diff
@@ -373,56 +369,61 @@ where
-    ///
-    /// We assume the linear data uses the same port offsets on both sides of the node.
-    /// In the future we may want to have a more general mechanism to handle this.
-    //
-    // Note that `Command::linear_units` assumes this behaviour.
-    fn follow_linear_port(&self, node: Node, port: impl Into<Port>) -> Option<Port> {
-        let port = port.into();
-        let optype = self.circ.get_optype(node);
-        if !optype.port_kind(port)?.is_linear() {
-            return None;
-        }
-        let other_port = Port::new(port.direction().reverse(), port.index());
-        if optype.port_kind(other_port) == optype.port_kind(port) {
-            Some(other_port)
-        } else {
-            None
+        let mut input_linear_units = Vec::new();
+        let mut output_linear_units = Vec::new();
+
+        let input_units = Units::new_incoming(self.circ, node, DefaultUnitLabeller)
+            .filter_map(filter::filter_linear);
+        let output_units = Units::new_outgoing(self.circ, node, DefaultUnitLabeller)
+            .filter_map(filter::filter_linear);
+        for ports in input_units.zip_longest(output_units) {
+            match ports {
+                EitherOrBoth::Right((_, out_port, _)) => {
+                    // Add a new linear unit for this output port.
+                    let new_wire = Wire::new(node, out_port);
+                    let linear_id = self.wire_unit.len();
+                    self.wire_unit.insert(new_wire, linear_id);
+                    output_linear_units.push(LinearUnit::new(linear_id));
+                }
+                EitherOrBoth::Left((_, in_port, _)) => {
+                    // Terminate the input linear unit.
+                    let Some(linear_id) = self.circ.linked_outputs(node, in_port).next().and_then(
+                        |(wire_node, wire_port)| {
+                            self.wire_unit.remove(&Wire::new(wire_node, wire_port))
+                        },
+                    ) else {
+                        continue;
+                    };
+                    input_linear_units.push(LinearUnit::new(linear_id));
+                }
+                EitherOrBoth::Both((_, in_port, _), (_, out_port, _)) => {
+                    // Update the input linear unit using the output port.
+                    let Some(linear_id) = self.circ.linked_outputs(node, in_port).next().and_then(
+                        |(wire_node, wire_port)| {
+                            self.wire_unit.remove(&Wire::new(wire_node, wire_port))
+                        },
+                    ) else {
+                        continue;
+                    };
+                    let linear_unit = LinearUnit::new(linear_id);
+                    let new_wire = Wire::new(node, out_port);
+                    input_linear_units.push(linear_unit);
+                    output_linear_units.push(linear_unit);
+                    self.wire_unit.insert(new_wire, linear_id);
```

Comment:
would be good to extract the logic in the previous cases in to functions and re-use them here